File: mobile/ios/Runner/Services/BridgeService.swift
import Foundation

/**
 * Bridge Service - iOS Native Implementation
 * Cross-chain bridging
 */
class BridgeService {
    static let API_BASE = "https://api.tigerwallet.com/api/v1"
    var authToken: String?
    
    init(token: String? = nil) {
        self.authToken = token
    }
    
    private func createRequest(endpoint: String, method: String = "GET", body: [String: Any]? = nil) -> URLRequest? {
        guard let url = URL(string: "\(Self.API_BASE)\(endpoint)") else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body = body {
            request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        }
        return request
    }
    
    /// Get supported chains
    func getSupportedChains(completion: @escaping (Result<[Chain], Error>) -> Void) {
        guard let request = createRequest(endpoint: "/bridge/chains") else {
            completion(.failure(BridgeError.invalidRequest))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, _, error in
            if let error = error { completion(.failure(error)); return }
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let chainsData = json["data"] as? [[String: Any]] else {
                completion(.failure(BridgeError.invalidResponse)); return
            }
            
            let chains = chainsData.compactMap { c -> Chain? in
                guard let id = c["id"] as? String, let name = c["name"] as? String else { return nil }
                return Chain(id: id, name: name, symbol: c["symbol"] as? String ?? "", isActive: c["isActive"] as? Bool ?? false)
            }
            completion(.success(chains))
        }.resume()
    }
    
    /// Get tokens for chain
    func getTokens(chain: String, completion: @escaping (Result<[BridgeToken], Error>) -> Void) {
        guard let request = createRequest(endpoint: "/bridge/tokens?chain=\(chain)") else {
            completion(.failure(BridgeError.invalidRequest)); return
        }
        
        URLSession.shared.dataTask(with: request) { data, _, error in
            if let error = error { completion(.failure(error)); return }
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let tokensData = json["data"] as? [[String: Any]] else {
                completion(.failure(BridgeError.invalidResponse)); return
            }
            
            let tokens = tokensData.compactMap { t -> BridgeToken? in
                guard let token = t["token"] as? String else { return nil }
                return BridgeToken(
                    token: token,
                    name: t["name"] as? String ?? "",
                    symbol: t["symbol"] as? String ?? "",
                    minAmount: t["minAmount"] as? Double ?? 0,
                    maxAmount: t["maxAmount"] as? Double ?? 0,
                    isActive: t["isActive"] as? Bool ?? false
                )
            }
            completion(.success(tokens))
        }.resume()
    }
    
    /// Get bridge estimate
    func getEstimate(fromChain: String, toChain: String, token: String, amount: Double, completion: @escaping (Result<BridgeEstimate, Error>) -> Void) {
        guard let request = createRequest(endpoint: "/bridge/estimate?fromChain=\(fromChain)&toChain=\(toChain)&token=\(token)&amount=\(amount)") else {
            completion(.failure(BridgeError.invalidRequest)); return
        }
        
        URLSession.shared.dataTask(with: request) { data, _, error in
            if let error = error { completion(.failure(error)); return }
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let estimateData = json["data"] as? [String: Any] else {
                completion(.failure(BridgeError.invalidResponse)); return
            }
            
            let estimate = BridgeEstimate(
                receivedAmount: estimateData["receivedAmount"] as? Double ?? 0,
                fee: estimateData["fee"] as? Double ?? 0,
                feePercentage: estimateData["feePercentage"] as? Double ?? 0,
                estimatedTime: estimateData["estimatedTime"] as? String ?? ""
            )
            completion(.success(estimate))
        }.resume()
    }
    
    /// Initiate bridge transaction
    func initiateBridge(fromChain: String, toChain: String, token: String, amount: Double, completion: @escaping (Result<BridgeTransaction, Error>) -> Void) {
        guard let request = createRequest(endpoint: "/bridge/transactions", method: "POST", body: [
            "fromChain": fromChain,
            "toChain": toChain,
            "token": token,
            "amount": amount
        ]) else {
            completion(.failure(BridgeError.invalidRequest)); return
        }
        
        URLSession.shared.dataTask(with: request) { data, _, error in
            if let error = error { completion(.failure(error)); return }
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let txData = json["data"] as? [String: Any] else {
                completion(.failure(BridgeError.invalidResponse)); return
            }
            
            let tx = BridgeTransaction(
                id: txData["id"] as? String ?? "",
                fromChain: txData["fromChain"] as? String ?? "",
                toChain: txData["toChain"] as? String ?? "",
                token: txData["token"] as? String ?? "",
                amount: txData["amount"] as? Double ?? 0,
                fee: txData["fee"] as? Double ?? 0,
                receivedAmount: txData["receivedAmount"] as? Double ?? 0,
                status: txData["status"] as? String ?? "PENDING"
            )
            completion(.success(tx))
        }.resume()
    }
    
    /// Get user transactions
    func getUserTransactions(completion: @escaping (Result<[BridgeTransaction], Error>) -> Void) {
        guard let request = createRequest(endpoint: "/bridge/transactions") else {
            completion(.failure(BridgeError.invalidRequest)); return
        }
        
        URLSession.shared.dataTask(with: request) { data, _, error in
            if let error = error { completion(.failure(error)); return }
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let txsData = json["data"] as? [[String: Any]] else {
                completion(.failure(BridgeError.invalidResponse)); return
            }
            
            let txs = txsData.compactMap { t -> BridgeTransaction? in
                guard let id = t["id"] as? String else { return nil }
                return BridgeTransaction(
                    id: id,
                    fromChain: t["fromChain"] as? String ?? "",
                    toChain: t["toChain"] as? String ?? "",
                    token: t["token"] as? String ?? "",
                    amount: t["amount"] as? Double ?? 0,
                    fee: t["fee"] as? Double ?? 0,
                    receivedAmount: t["receivedAmount"] as? Double ?? 0,
                    status: t["status"] as? String ?? "PENDING"
                )
            }
            completion(.success(txs))
        }.resume()
    }
}

struct Chain {
    let id: String
    let name: String
    let symbol: String
    let isActive: Bool
}

struct BridgeToken {
    let token: String
    let name: String
    let symbol: String
    let minAmount: Double
    let maxAmount: Double
    let isActive: Bool
}

struct BridgeEstimate {
    let receivedAmount: Double
    let fee: Double
    let feePercentage: Double
    let estimatedTime: String
}

struct BridgeTransaction {
    let id: String
    let fromChain: String
    let toChain: String
    let token: String
    let amount: Double
    let fee: Double
    let receivedAmount: Double
    let status: String
}

enum BridgeError: Error {
    case invalidRequest
    case invalidResponse
}

File: mobile_apps/android_legacy/app/src/main/java/com/tigerwallet/app/services/BridgeService.java
package com.tigerwallet.app.services;

import org.json.JSONArray;
import org.json.JSONObject;
import java.net.HttpURLConnection;
import java.net.URL;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

/**
 * Bridge Service - Android Native Implementation
 */
public class BridgeService {
    private static final String API_BASE = "https://api.tigerwallet.com/api/v1";
    private String authToken;
    
    public BridgeService(String token) { this.authToken = token; }
    
    private HttpURLConnection createConnection(String endpoint) throws Exception {
        URL url = new URL(API_BASE + endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        return conn;
    }
    
    public Chain[] getSupportedChains() throws Exception {
        HttpURLConnection conn = createConnection("/bridge/chains");
        if (conn.getResponseCode() == 200) {
            String response = readResponse(conn);
            JSONArray data = new JSONObject(response).getJSONArray("data");
            Chain[] chains = new Chain[data.length()];
            for (int i = 0; i < data.length(); i++) {
                JSONObject c = data.getJSONObject(i);
                chains[i] = new Chain(c.getString("id"), c.getString("name"), c.getString("symbol"), c.getBoolean("isActive"));
            }
            return chains;
        }
        return new Chain[0];
    }
    
    public BridgeToken[] getTokens(String chain) throws Exception {
        HttpURLConnection conn = createConnection("/bridge/tokens?chain=" + chain);
        if (conn.getResponseCode() == 200) {
            String response = readResponse(conn);
            JSONArray data = new JSONObject(response).getJSONArray("data");
            BridgeToken[] tokens = new BridgeToken[data.length()];
            for (int i = 0; i < data.length(); i++) {
                JSONObject t = data.getJSONObject(i);
                tokens[i] = new BridgeToken(t.getString("token"), t.getString("name"), t.getDouble("minAmount"), t.getDouble("maxAmount"), t.getBoolean("isActive"));
            }
            return tokens;
        }
        return new BridgeToken[0];
    }
    
    public BridgeEstimate getEstimate(String fromChain, String toChain, String token, double amount) throws Exception {
        HttpURLConnection conn = createConnection("/bridge/estimate?fromChain=" + fromChain + "&toChain=" + toChain + "&token=" + token + "&amount=" + amount);
        if (conn.getResponseCode() == 200) {
            String response = readResponse(conn);
            JSONObject data = new JSONObject(response).getJSONObject("data");
            return new BridgeEstimate(data.getDouble("receivedAmount"), data.getDouble("fee"), data.getDouble("feePercentage"), data.getString("estimatedTime"));
        }
        return null;
    }
    
    public BridgeTransaction initiateBridge(String fromChain, String toChain, String token, double amount) throws Exception {
        URL url = new URL(API_BASE + "/bridge/transactions");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        conn.setDoOutput(true);
        
        JSONObject body = new JSONObject();
        body.put("fromChain", fromChain);
        body.put("toChain", toChain);
        body.put("token", token);
        body.put("amount", amount);
        conn.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        
        if (conn.getResponseCode() == 201) {
            String response = readResponse(conn);
            JSONObject data = new JSONObject(response).getJSONObject("data");
            return new BridgeTransaction(data.getString("id"), data.getString("fromChain"), data.getString("toChain"), 
                data.getString("token"), data.getDouble("amount"), data.getDouble("fee"), data.getDouble("receivedAmount"), data.getString("status"));
        }
        return null;
    }
    
    private String readResponse(HttpURLConnection conn) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) response.append(line);
        reader.close();
        return response.toString();
    }
    
    public static class Chain {
        public String id, name, symbol;
        public boolean isActive;
        public Chain(String id, String name, String symbol, boolean isActive) {
            this.id = id; this.name = name; this.symbol = symbol; this.isActive = isActive;
        }
    }
    
    public static class BridgeToken {
        public String token, name;
        public double minAmount, maxAmount;
        public boolean isActive;
        public BridgeToken(String token, String name, double minAmount, double maxAmount, boolean isActive) {
            this.token = token; this.name = name; this.minAmount = minAmount; this.maxAmount = maxAmount; this.isActive = isActive;
        }
    }
    
    public static class BridgeEstimate {
        public double receivedAmount, fee, feePercentage;
        public String estimatedTime;
        public BridgeEstimate(double receivedAmount, double fee, double feePercentage, String estimatedTime) {
            this.receivedAmount = receivedAmount; this.fee = fee; this.feePercentage = feePercentage; this.estimatedTime = estimatedTime;
        }
    }
    
    public static class BridgeTransaction {
        public String id, fromChain, toChain, token, status;
        public double amount, fee, receivedAmount;
        public BridgeTransaction(String id, String fromChain, String toChain, String token, double amount, double fee, double receivedAmount, String status) {
            this.id = id; this.fromChain = fromChain; this.toChain = toChain; this.token = token;
            this.amount = amount; this.fee = fee; this.receivedAmount = receivedAmount; this.status = status;
        }
    }
}
